<div style="text-align: center;">
  <img src="https://github.com/Hack-io-Data/Imagenes/blob/main/01-LogosHackio/logo_naranja@4x.png?raw=true" alt="esquema" />
</div>

# Laboratorio Pandas (EDA y Unión de Datos)

## Los datos

Para este laboratorio, trabajarás con los siguientes conjuntos de datos:

- **`netflix_originals.csv`**: Contiene información sobre producciones originales de Netflix. Las columnas que nos encontraremos en este conjunto de datos son: 

   - `Title`: Contiene el nombre del título de la producción original de Netflix.

   - `Genre`: Especifica el género o combinación de géneros de la producción.

   - `Premiere`: Indica la fecha de estreno de la producción en Netflix.

   - `Runtime`: Contiene la duración de la producción en minutos.

   - `IMDB Score`: Representa la calificación promedio que la producción ha recibido en la plataforma IMDB.

   - `Language`: Especifica el idioma principal o los idiomas en los que se presenta la producción.

- **`netflix_titles.csv`**: Contiene información sobre una variedad de títulos disponibles en la plataforma Netflix, incluyendo tanto películas como series de televisión. Las columnas que nos encontraremos en este conjunto de datos son:

   - `show_id`: Identificador único para cada título en el dataset.

   - `type`: Especifica el tipo de contenido, ya sea "Movie" (película) o "TV Show" (serie de televisión).

   - `title`: Contiene el nombre del título de la producción.

   - `director`: Muestra el nombre del director del título. Puede estar vacío en algunos casos.

   - `cast`: Lista de actores y actrices que participaron en el título. También puede estar vacío en algunos casos.

   - `country`: País o países donde se produjo el título. Esta columna también puede tener valores faltantes.

   - `date_added`: Fecha en la que el título fue añadido al catálogo de Netflix.

   - `release_year`: Año en que el título fue lanzado o producido.

   - `rating`: Clasificación del contenido en términos de audiencia (por ejemplo, PG-13, TV-MA).

   - `duration`: Duración del título. Para películas, se expresa en minutos (e.g., "90 min") y para series en temporadas (e.g., "2 Seasons").

   - `listed_in`: Categorías o géneros bajo los cuales el título está clasificado (por ejemplo, "Documentaries", "TV Dramas").

   - `description`: Breve sinopsis o descripción del contenido del título.

A lo largo de los laboratorios de Pandas, trabajarás con estos conjuntos de datos (o con los derivados del Laboratorio), algunas de las actividades que realizaremos son:

- Exploración de datos y análisis exploratorio (EDA) para familiarizarse con los conjuntos de datos.

- Realización de uniones entre conjuntos de datos, aplicando diferentes técnicas de *merging* y analizando la relevancia de cada método.

- Uso de funciones de selección y filtrado (`loc`, `iloc`) para extraer información clave.

- Desarrollo de análisis descriptivos y visualización de patrones para comprender mejor las decisiones estratégicas de la plataforma.

----

## Objetivo General de la Serie de Laboratorios



1. **Unión de `netflix_titles.csv` con `netflix_originals.csv`:**

   - **Análisis de contenido original:**  Identificarás los títulos que son producciones originales de Netflix y los compararás con los títulos adquiridos. Este análisis te permitirá:

     - Comparar la diversidad de géneros entre los títulos originales y no originales.

     - Evaluar si ciertos géneros o tipos de contenido (como documentales o series) son más frecuentes en las producciones originales.

     - Investigar si la producción original se concentra en determinados países o idiomas.


   - **Evaluación de contenido:** En esta etapa, explorarás si las producciones originales de Netflix tienden a recibir mejores o peores evaluaciones en comparación con el contenido adquirido. Algunas preguntas que se buscarán responder incluyen:

     - ¿Los títulos originales tienen una calificación promedio superior en plataformas como IMDB?

     - ¿Son las producciones originales más consistentes en términos de calidad percibida?

     - ¿Existen diferencias en la popularidad de los títulos originales según su género o país de origen?


   - **Identificación de tendencias:**  Por último, este análisis te permitirá explorar cómo ha evolucionado la estrategia de contenido de Netflix a lo largo del tiempo. Algunos puntos clave a investigar serán:

     - ¿Cómo ha crecido la producción original en comparación con la adquisición de contenido?

     - ¿Qué géneros o tipos de contenido han recibido mayor inversión en los últimos años?
     
     - ¿Existen patrones en los países o idiomas de las producciones originales, especialmente en relación con la expansión global de Netflix?


----


## Ejercicio: Análisis y Enriquecimiento del Catálogo de Netflix

### Parte 1: Análisis Exploratorio de Datos (EDA)

1. **Carga de los datos:**

   - Deberás cargar los dos conjuntos de datos que te hemos proporcionado para realizar este Laboratorio:

     - `netflix_originals.csv`

     - `netflix_titles.csv`

   Deberás revisar la estructura de cada conjunto de datos utilizando los métodos aprendidos durante la lección.

2. **Análisis de columnas y consistencia:**

   - Deberás hacer un análisis exploratorio de cada uno de los conjuntos de datos. Será obligatorio que incluyas las conclusiones generales obtenidas de cada uno de los análisis exploratorios en el Jupyter Notebook de tus soluciones.

   - Explorar qué columnas contiene cada conjunto de datos y cuáles son sus tipos de datos. Identificar si hay valores nulos o duplicados.

   - Algunas preguntas que nos podemos plantear en esta fase son (recuerda que esto es solo un ejemplo, y que podemos explorar todo lo que queramos/necesitemos para entender lo mejor posible los conjuntos de datos):

      - ¿Cuáles son los tipos de datos de cada columna? Nos puede ayudar a detectar si algún tipo de dato está incorrectamente asignado, como fechas almacenadas como cadenas de texto.

      - ¿Existen valores nulos? ¿En qué columnas están? Este paso es crucial para decidir si deben ser eliminados, rellenados o ignorados.

      - ¿Cuántos títulos únicos hay en cada conjunto de datos? Nos puede permitir verificar si existen títulos duplicados que puedan distorsionar el análisis.

      - ¿Qué géneros tenemos en el conjunto de datos? Esto nos puede ayudar a identificar los géneros más populares o incluso a identificar una posible limpieza de dicha columna para poder extraer conclusiones más claras en nuestro análisis.

### Parte 2: Unión de Datos

1. **Realización de la unión:**

   - Deberás realizar la unión que consideres más apropiada entre `netflix_titles.csv` y `netflix_originals.csv`. Además, deberás justificar el tipo de unión que has realizado.

   - Esto nos ayudará a contestar preguntas más adelente del tipo:

     - ¿Cuántos títulos en `netflix_titles.csv` son producciones originales de Netflix?

     - ¿Hay diferencias significativas en la puntuación de IMDB entre títulos originales y no originales?

### Parte 3: Uso de `loc` e `iloc`

Usando el DataFrame creado en la Parte 2: 

1. **Filtrado de datos con `loc`:**

   - Selecciona todas las filas donde el tipo de contenido sea “Movie”.

   - Muestra solo los títulos (title) y la duración (duration) de todos los contenidos que sean de tipo “TV Show”.

   - Selecciona todas las filas donde el país sea “United States”.

   - Muestra todas las películas (type = “Movie”) y selecciona las columnas title y director.

   - Selecciona los títulos (title) y géneros (listed_in) de todas las producciones lanzadas en 2018.

   - Selecciona las filas donde la columna director esté vacía y muestra solo los títulos (title).

2. **Acceso a datos específicos con `iloc`:**

   - Selecciona las primeras 5 filas del conjunto de datos y muestra solo las columnas title, director y country.

   - Muestra las últimas 5 filas del conjunto de datos y selecciona las columnas title y description.

   - Selecciona las primeras 10 filas del dataset y muestra las columnas desde la segunda hasta la quinta (incluidas). 

   - Muestra las últimas 7 filas del dataset y selecciona las columnas title, director, y country usando índices posicionales.

   - Selecciona las filas desde la 15 hasta la 25 (incluidas) y muestra las primeras 4 columnas.

   - Muestra las filas 20 a 30 y solo las columnas title, release_year, y rating usando índices posicionales.


**NOTA IMPORTANTE**: Antes de terminar el lab, guarda el conjunto de datos creado en la fase de unión para trabajar con el en las siguientes lecciones. 

In [2]:
import pandas as pd
import numpy as np

pd.set_option('display.max_columns', None) #hace que jupyter muestre todas las columnas

In [16]:
df_netflix = pd.read_csv("../datos/netflix_originals.csv")
df_netflix_titles = pd.read_csv("../datos/netflix_titles.csv")

In [17]:
df_netflix.head()

Unnamed: 0            Title                  Genre           Premiere  \
0           0  Enter the Anime            Documentary     August 5, 2019   
1           1      Dark Forces               Thriller    August 21, 2020   
2           2          The App  Science fiction/Drama  December 26, 2019   
3           3   The Open House        Horror thriller   January 19, 2018   
4           4      Kaali Khuhi                Mystery   October 30, 2020   

   Runtime  IMDB Score          Language  
0       58         2.5  English/Japanese  
1       81         2.6           Spanish  
2       79         2.6           Italian  
3       94         3.2           English  
4       90         3.4             Hindi

In [18]:
df_netflix.drop(columns = ["Unnamed: 0"], inplace = True) 
# elimina primera columna "Unnamed" no aporta valor
# inplace guarda el cambio

In [19]:
df_netflix.head()

Title                  Genre           Premiere  Runtime  \
0  Enter the Anime            Documentary     August 5, 2019       58   
1      Dark Forces               Thriller    August 21, 2020       81   
2          The App  Science fiction/Drama  December 26, 2019       79   
3   The Open House        Horror thriller   January 19, 2018       94   
4      Kaali Khuhi                Mystery   October 30, 2020       90   

   IMDB Score          Language  
0         2.5  English/Japanese  
1         2.6           Spanish  
2         2.6           Italian  
3         3.2           English  
4         3.4             Hindi

In [23]:
df_netflix_titles.head()

Unnamed: 0 show_id     type                  title         director  \
0           0      s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1           1      s2  TV Show          Blood & Water              NaN   
2           2      s3  TV Show              Ganglands  Julien Leclercq   
3           3      s4  TV Show  Jailbirds New Orleans              NaN   
4           4      s5  TV Show           Kota Factory              NaN   

                                                cast        country  \
0                                                NaN  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2  Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
3                                                NaN            NaN   
4  Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   

           date_added  release_year rating   duration  \
0  September 25, 2021          2020  PG-13     90 min   
1  September 24, 2021          2021  TV-MA  2 Seasons   
2  September 24, 2021          2021  TV-MA        NaN   
3  September 24, 2021          2021  TV-MA        NaN   
4  September 24, 2021          2021  TV-MA  2 Seasons   

                                           listed_in  \
0                                      Documentaries   
1    International TV Shows, TV Dramas, TV Mysteries   
2  Crime TV Shows, International TV Shows, TV Act...   
3                             Docuseries, Reality TV   
4  International TV Shows, Romantic TV Shows, TV ...   

                                         description  
0  As her father nears the end of his life, filmm...  
1  After crossing paths at a party, a Cape Town t...  
2  To protect his family from a powerful drug lor...  
3  Feuds, flirtations and toilet talk go down amo...  
4  In a city of coaching centers known to train I...

In [24]:
df_netflix_titles.drop(columns = ["Unnamed: 0"], inplace = True)
#Columna "unnamed" no aporta valor

In [26]:
df_netflix_titles.head()

show_id     type                  title         director  \
0      s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1      s2  TV Show          Blood & Water              NaN   
2      s3  TV Show              Ganglands  Julien Leclercq   
3      s4  TV Show  Jailbirds New Orleans              NaN   
4      s5  TV Show           Kota Factory              NaN   

                                                cast        country  \
0                                                NaN  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2  Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
3                                                NaN            NaN   
4  Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   

           date_added  release_year rating   duration  \
0  September 25, 2021          2020  PG-13     90 min   
1  September 24, 2021          2021  TV-MA  2 Seasons   
2  September 24, 2021          2021  TV-MA        NaN   
3  September 24, 2021          2021  TV-MA        NaN   
4  September 24, 2021          2021  TV-MA  2 Seasons   

                                           listed_in  \
0                                      Documentaries   
1    International TV Shows, TV Dramas, TV Mysteries   
2  Crime TV Shows, International TV Shows, TV Act...   
3                             Docuseries, Reality TV   
4  International TV Shows, Romantic TV Shows, TV ...   

                                         description  
0  As her father nears the end of his life, filmm...  
1  After crossing paths at a party, a Cape Town t...  
2  To protect his family from a powerful drug lor...  
3  Feuds, flirtations and toilet talk go down amo...  
4  In a city of coaching centers known to train I...

In [30]:
#Veo la información que contiene cada data_set
df_netflix.columns


Index(['Title', 'Genre', 'Premiere', 'Runtime', 'IMDB Score', 'Language'], dtype='object')

In [31]:
df_netflix_titles.columns

Index(['show_id', 'type', 'title', 'director', 'cast', 'country', 'date_added',
       'release_year', 'rating', 'duration', 'listed_in', 'description'],
      dtype='object')

### Tengo 2 dataframes:
1. El primero contiene id, título, género, premiere, duración, imdb e idioma.
2. El segundo contiene id, tipo, título, director, reparto, país, día añadido, año lanzamiento, rating, duración, listed_in y descripción.
3. De ambos he eliminado la columna "Unnamed: 0" por no aportar valor.
4. En común tenemos el título y la duración.

In [38]:
# Busco títulos duplicados, porque no debería haberlos duplicados 
df_netflix.duplicated(subset = "Title").sum()
# Resultado 0 ok


np.int64(0)

In [41]:
# Busco títulos y IDs duplicados, porque no debería haberlos duplicados 
df_netflix_titles.duplicated(subset = ["title", "show_id"]).sum()
# Resultado 0 ok


np.int64(0)

In [84]:
#Hago un merge de las dos tablas por el dato que tienen en común, que en este caso es el título, pero las columnas tienen distinto nombre en los dos data frames.

df_merge = df_netflix_titles.merge(df_netflix, left_on = "title", right_on = "Title", how = "left")
df_merge


show_id     type                  title         director  \
0         s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1         s2  TV Show          Blood & Water              NaN   
2         s3  TV Show              Ganglands  Julien Leclercq   
3         s4  TV Show  Jailbirds New Orleans              NaN   
4         s5  TV Show           Kota Factory              NaN   
...      ...      ...                    ...              ...   
8802   s8803    Movie                 Zodiac    David Fincher   
8803   s8804  TV Show            Zombie Dumb              NaN   
8804   s8805    Movie             Zombieland  Ruben Fleischer   
8805   s8806    Movie                   Zoom     Peter Hewitt   
8806   s8807    Movie                 Zubaan      Mozez Singh   

                                                   cast        country  \
0                                                   NaN  United States   
1     Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2     Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
3                                                   NaN            NaN   
4     Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   
...                                                 ...            ...   
8802  Mark Ruffalo, Jake Gyllenhaal, Robert Downey J...  United States   
8803                                                NaN            NaN   
8804  Jesse Eisenberg, Woody Harrelson, Emma Stone, ...  United States   
8805  Tim Allen, Courteney Cox, Chevy Chase, Kate Ma...  United States   
8806  Vicky Kaushal, Sarah-Jane Dias, Raaghav Chanan...          India   

              date_added  release_year rating   duration  \
0     September 25, 2021          2020  PG-13     90 min   
1     September 24, 2021          2021  TV-MA  2 Seasons   
2     September 24, 2021          2021  TV-MA        NaN   
3     September 24, 2021          2021  TV-MA        NaN   
4     September 24, 2021          2021  TV-MA  2 Seasons   
...                  ...           ...    ...        ...   
8802   November 20, 2019          2007      R    158 min   
8803        July 1, 2019          2018  TV-Y7  2 Seasons   
8804    November 1, 2019          2009      R        NaN   
8805    January 11, 2020          2006     PG        NaN   
8806       March 2, 2019          2015  TV-14    111 min   

                                              listed_in  \
0                                         Documentaries   
1       International TV Shows, TV Dramas, TV Mysteries   
2     Crime TV Shows, International TV Shows, TV Act...   
3                                Docuseries, Reality TV   
4     International TV Shows, Romantic TV Shows, TV ...   
...                                                 ...   
8802                     Cult Movies, Dramas, Thrillers   
8803             Kids' TV, Korean TV Shows, TV Comedies   
8804                            Comedies, Horror Movies   
8805                 Children & Family Movies, Comedies   
8806     Dramas, International Movies, Music & Musicals   

                                            description                 Title  \
0     As her father nears the end of his life, filmm...  Dick Johnson Is Dead   
1     After crossing paths at a party, a Cape Town t...                   NaN   
2     To protect his family from a powerful drug lor...                   NaN   
3     Feuds, flirtations and toilet talk go down amo...                   NaN   
4     In a city of coaching centers known to train I...                   NaN   
...                                                 ...                   ...   
8802  A political cartoonist, a crime reporter and a...                   NaN   
8803  While living alone in a spooky town, a young g...                   NaN   
8804  Looking to survive in a world taken over by zo...                   NaN   
8805  Dragged from civilian life, a former superhero...                   NaN   
8806  A scrapp

In [ ]:
# La columna duration tiene datos equivocados, por ejemplo, "2 seasons", además, el formato es de tipo string, i.e. "90 min", lo que hace imposible aplicar operaciones matemáticas.

In [ ]:
#Columna "Title" ahora se encuentra repetida, así que la eliminamos.
df_merge.drop(columns = ["Title"], inplace = True)


In [106]:
# Cuento cantidad de títulos originales que encuentro en el DataFrame de Originiales
# Contando cantidad de NO nulos para cada columna
df_merge.shape[0] - df_merge.isnull().sum()

show_id         8807
type            8807
title           8807
director        6173
cast            7982
country         7976
date_added      8797
release_year    8807
rating          8803
duration        3994
listed_in       8807
description     8807
Title            513
Genre            513
Premiere         513
Runtime          513
IMDB Score       513
Language         513
dtype: int64

Hay 513 títulos originales de Netflix.

No se puede determinar si hay diferencias significativas entre la puntuación IMDB de las producciones originales de Netflix y las que no lo son, puesto que no tenemos la puntuación para aquellas que no lo son.

In [85]:
# Reviso tipos de datos 
df_merge.dtypes

show_id          object
type             object
title            object
director         object
cast             object
country          object
date_added       object
release_year      int64
rating           object
duration         object
listed_in        object
description      object
Title            object
Genre            object
Premiere         object
Runtime         float64
IMDB Score      float64
Language         object
dtype: object

In [ ]:
# El formato de duration no es correcto.
# El tipo de la fecha debería ser int para poder operar sobre ella.
# Misma situación con la duración, debería ser int.

In [88]:
# Busco los valores únicos de las distintas columnas:
for column in df_merge.select_dtypes("object").columns:
    print(column.upper())
    print(df_merge[column].unique())
    print("\n------------\n")

SHOW_ID
['s1' 's2' 's3' ... 's8805' 's8806' 's8807']

------------

TYPE
['Movie' 'TV Show']

------------

TITLE
['Dick Johnson Is Dead' 'Blood & Water' 'Ganglands' ... 'Zombieland'
 'Zoom' 'Zubaan']

------------

DIRECTOR
['Kirsten Johnson' nan 'Julien Leclercq' ... 'Majid Al Ansari'
 'Peter Hewitt' 'Mozez Singh']

------------

CAST
[nan
 'Ama Qamata, Khosi Ngema, Gail Mabalane, Thabang Molaba, Dillon Windvogel, Natasha Thahane, Arno Greeff, Xolile Tshabalala, Getmore Sithole, Cindy Mahlangu, Ryle De Morny, Greteli Fincham, Sello Maake Ka-Ncube, Odwa Gwanya, Mekaila Mathys, Sandi Schultz, Duane Williams, Shamilla Miller, Patrick Mofokeng'
 'Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabiha Akkari, Sofia Lesaffre, Salim Kechiouche, Noureddine Farihi, Geert Van Rampelberg, Bakary Diombera'
 ...
 'Jesse Eisenberg, Woody Harrelson, Emma Stone, Abigail Breslin, Amber Heard, Bill Murray, Derek Graf'
 'Tim Allen, Courteney Cox, Chevy Chase, Kate Mara, Ryan Newman, Michael Cassidy, Spencer 

In [ ]:
#Hay valores nulos en directores, reparto, países y rating. 

In [114]:
# Busco valores únicos

for column in df_merge.select_dtypes("object").columns:
    print(column.upper())
    print(df_merge[column].value_counts())
    print("\n------------\n")

SHOW_ID
show_id
s1       1
s5875    1
s5869    1
s5870    1
s5871    1
        ..
s2931    1
s2930    1
s2929    1
s2928    1
s8807    1
Name: count, Length: 8807, dtype: int64

------------

TYPE
type
Movie      6131
TV Show    2676
Name: count, dtype: int64

------------

TITLE
title
Dick Johnson Is Dead                     1
Ip Man 2                                 1
Hannibal Buress: Comedy Camisado         1
Turbo FAST                               1
Masha's Tales                            1
                                        ..
Love for Sale 2                          1
ROAD TO ROMA                             1
Good Time                                1
Captain Underpants Epic Choice-o-Rama    1
Zubaan                                   1
Name: count, Length: 8807, dtype: int64

------------

DIRECTOR
director
Rajiv Chilaka                     19
Raúl Campos, Jan Suter            18
Marcus Raboy                      16
Suhas Kadav                       16
Jay Karas          

In [ ]:
# En la columna duración, hay 425 filas que no corresponden con este dato, sino que es "2 Seasons", que representa un 10% de los datos.

# En la columna ratings hay 3 valores que se corresponden con la duración, y no el rating, pero son poco relevante porque representan un 0.03% del total.

In [91]:
for column in df_merge.select_dtypes("object").columns:
    print(column.upper())
    print(df_merge[column].value_counts(normalize = True)*100) #esto no tiene en cuenta los nulos
    print("\n------------\n")

SHOW_ID
show_id
s1       0.011355
s5875    0.011355
s5869    0.011355
s5870    0.011355
s5871    0.011355
           ...   
s2931    0.011355
s2930    0.011355
s2929    0.011355
s2928    0.011355
s8807    0.011355
Name: proportion, Length: 8807, dtype: float64

------------

TYPE
type
Movie      69.615079
TV Show    30.384921
Name: proportion, dtype: float64

------------

TITLE
title
Dick Johnson Is Dead                     0.011355
Ip Man 2                                 0.011355
Hannibal Buress: Comedy Camisado         0.011355
Turbo FAST                               0.011355
Masha's Tales                            0.011355
                                           ...   
Love for Sale 2                          0.011355
ROAD TO ROMA                             0.011355
Good Time                                0.011355
Captain Underpants Epic Choice-o-Rama    0.011355
Zubaan                                   0.011355
Name: proportion, Length: 8807, dtype: float64

------------



In [ ]:
# Los tipos más habituales de contenido en Netflix son películas (un 70%) y series de TV (30%).
# El 36% tienes un rating para mayores, y un 25% son aptas para menores de 14 años.

# Los géneros más comunes son:
# Documentales: 26% y Drama: 14%.

# El 69% del contenido está en inglés.

In [92]:
df_netflix_titles.shape

(8807, 12)

In [93]:
df_netflix.shape

(513, 6)

In [94]:
df_merge.shape

(8807, 18)

In [107]:
df_merge.describe(include = "O").T

count unique                                                top  \
show_id      8807   8807                                                 s1   
type         8807      2                                              Movie   
title        8807   8807                               Dick Johnson Is Dead   
director     6173   4528                                      Rajiv Chilaka   
cast         7982   7692                                 David Attenborough   
country      7976    748                                      United States   
date_added   8797   1767                                    January 1, 2020   
rating       8803     17                                              TV-MA   
duration     3994    185                                          2 Seasons   
listed_in    8807    514                       Dramas, International Movies   
description  8807   8775  Paranormal activity at a lush, abandoned prope...   
Title         513    513                               Dick Johnson Is Dead   
Genre         513    106                                        Documentary   
Premiere      513    350                                    October 2, 2020   
Language      513     37                                            English   

             freq  
show_id         1  
type         6131  
title           1  
director       19  
cast           19  
country      2818  
date_added    109  
rating       3207  
duration      425  
listed_in     362  
description     4  
Title           1  
Genre         132  
Premiere        5  
Language      352

In [ ]:
#Habiendo 8807 registros únicos del show id, que es el tamaño del dataframe, quiere decir que son válidos como identificadores únicos de cada fila.

In [109]:
# Contar cantidad de Nulos para cada columna
df_merge.isnull().sum()

np.round(df_merge.isnull().sum() / df_merge.shape[0]*100, 2)

show_id          0.00
type             0.00
title            0.00
director        29.91
cast             9.37
country          9.44
date_added       0.11
release_year     0.00
rating           0.05
duration        54.65
listed_in        0.00
description      0.00
Title           94.18
Genre           94.18
Premiere        94.18
Runtime         94.18
IMDB Score      94.18
Language        94.18
dtype: float64

In [ ]:
# Hay muchos valores nulos.
#Hay muchas columnas en las que faltan datos. Particularmente llama la atención la duración, donde falta casi más de la mitad de los registros, siendo un rato importante por ejemplo, para los algoritmos de recomendación al usuario.

# Ejercicios

In [117]:
#1.1 Selecciono todas las filas con el contenido Movie
df_merge.loc[df_merge['type'] == 'Movie']

show_id   type                             title  \
0         s1  Movie              Dick Johnson Is Dead   
6         s7  Movie  My Little Pony: A New Generation   
7         s8  Movie                           Sankofa   
9        s10  Movie                      The Starling   
12       s13  Movie                      Je Suis Karl   
...      ...    ...                               ...   
8801   s8802  Movie                           Zinzana   
8802   s8803  Movie                            Zodiac   
8804   s8805  Movie                        Zombieland   
8805   s8806  Movie                              Zoom   
8806   s8807  Movie                            Zubaan   

                           director  \
0                   Kirsten Johnson   
6     Robert Cullen, José Luis Ucha   
7                      Haile Gerima   
9                    Theodore Melfi   
12              Christian Schwochow   
...                             ...   
8801                Majid Al Ansari   
8802                  David Fincher   
8804                Ruben Fleischer   
8805                   Peter Hewitt   
8806                    Mozez Singh   

                                                   cast  \
0                                                   NaN   
6     Vanessa Hudgens, Kimiko Glenn, James Marsden, ...   
7     Kofi Ghanaba, Oyafunmike Ogunlano, Alexandra D...   
9     Melissa McCarthy, Chris O'Dowd, Kevin Kline, T...   
12    Luna Wedler, Jannis Niewöhner, Milan Peschel, ...   
...                                                 ...   
8801  Ali Suliman, Saleh Bakri, Yasa, Ali Al-Jabri, ...   
8802  Mark Ruffalo, Jake Gyllenhaal, Robert Downey J...   
8804  Jesse Eisenberg, Woody Harrelson, Emma Stone, ...   
8805  Tim Allen, Courteney Cox, Chevy Chase, Kate Ma...   
8806  Vicky Kaushal, Sarah-Jane Dias, Raaghav Chanan...   

                                                country          date_added  \
0                                         United States  September 25, 2021   
6                                                   NaN  September 24, 2021   
7     United States, Ghana, Burkina Faso, United Kin...  September 24, 2021   
9                                         United States  September 24, 2021   
12                              Germany, Czech Republic  September 23, 2021   
...                                                 ...                 ...   
8801                       United Arab Emirates, Jordan       March 9, 2016   
8802                                      United States   November 20, 2019   
8804                                      United States    November 1, 2019   
8805                                      United States    January 11, 2020   
8806                                              India       March 2, 2019   

      release_year rating duration  \
0             2020  PG-13   90 min   
6             2021     PG      NaN   
7             1993  TV-MA  125 min   
9             2021  PG-13  104 min   
12            2021  TV-MA  127 min   
...            ...    ...      ...   
8801          2015  TV-MA      NaN   
8802          2007      R  158 min   
8804          2009      R      NaN   
8805          2006     PG      NaN   
8806          2015  TV-14  111 min   

                                             listed_in  \
0                                        Documentaries   
6                             Children & Family Movies   
7     Dramas, Independent Movies, International Movies   
9                                     Comedies, Dramas   
12                        Dramas, International Movies   
...                                                ...   
8801           Dramas, International Movies, Thrillers   
8802                    Cult Movies, Dramas, Thrillers   
8804                           Comedies, Horror Movies   
8805                Children & Family Movies, Comedies   
8806    Dramas, International Movies, Music & Musicals   

                                       

In [123]:
# 1.2 Títulos y duración de todos los contenidos que sean "TV show"
df_merge.loc[df_merge["type"] == 'TV Show', ["title", "duration"]]

title   duration
1             Blood & Water  2 Seasons
2                 Ganglands        NaN
3     Jailbirds New Orleans        NaN
4              Kota Factory  2 Seasons
5             Midnight Mass        NaN
...                     ...        ...
8795        Yu-Gi-Oh! Arc-V  2 Seasons
8796             Yunus Emre  2 Seasons
8797              Zak Storm        NaN
8800     Zindagi Gulzar Hai        NaN
8803            Zombie Dumb  2 Seasons

[2676 rows x 2 columns]

In [119]:
# 1.3 Todas las filas donde el país sea "TV show"
df_merge.loc[df_merge['country'] == 'United States']

show_id     type                 title         director  \
0         s1    Movie  Dick Johnson Is Dead  Kirsten Johnson   
9        s10    Movie          The Starling   Theodore Melfi   
15       s16  TV Show     Dear White People              NaN   
27       s28    Movie             Grown Ups     Dennis Dugan   
28       s29    Movie            Dark Skies    Scott Stewart   
...      ...      ...                   ...              ...   
8791   s8792    Movie           Young Adult    Jason Reitman   
8793   s8794    Movie  Yours, Mine and Ours     Raja Gosnell   
8802   s8803    Movie                Zodiac    David Fincher   
8804   s8805    Movie            Zombieland  Ruben Fleischer   
8805   s8806    Movie                  Zoom     Peter Hewitt   

                                                   cast        country  \
0                                                   NaN  United States   
9     Melissa McCarthy, Chris O'Dowd, Kevin Kline, T...  United States   
15    Logan Browning, Brandon P. Bell, DeRon Horton,...  United States   
27    Adam Sandler, Kevin James, Chris Rock, David S...  United States   
28    Keri Russell, Josh Hamilton, J.K. Simmons, Dak...  United States   
...                                                 ...            ...   
8791  Charlize Theron, Patton Oswalt, Patrick Wilson...  United States   
8793  Dennis Quaid, Rene Russo, Sean Faris, Katija P...  United States   
8802  Mark Ruffalo, Jake Gyllenhaal, Robert Downey J...  United States   
8804  Jesse Eisenberg, Woody Harrelson, Emma Stone, ...  United States   
8805  Tim Allen, Courteney Cox, Chevy Chase, Kate Ma...  United States   

              date_added  release_year rating duration  \
0     September 25, 2021          2020  PG-13   90 min   
9     September 24, 2021          2021  PG-13  104 min   
15    September 22, 2021          2021  TV-MA      NaN   
27    September 20, 2021          2010  PG-13      NaN   
28    September 19, 2021          2013  PG-13      NaN   
...                  ...           ...    ...      ...   
8791   November 20, 2019          2011      R      NaN   
8793   November 20, 2019          2005     PG      NaN   
8802   November 20, 2019          2007      R  158 min   
8804    November 1, 2019          2009      R      NaN   
8805    January 11, 2020          2006     PG      NaN   

                                 listed_in  \
0                            Documentaries   
9                         Comedies, Dramas   
15                  TV Comedies, TV Dramas   
27                                Comedies   
28         Horror Movies, Sci-Fi & Fantasy   
...                                    ...   
8791  Comedies, Dramas, Independent Movies   
8793    Children & Family Movies, Comedies   
8802        Cult Movies, Dramas, Thrillers   
8804               Comedies, Horror Movies   
8805    Children & Family Movies, Comedies   

                                            description                 Title  \
0     As her father nears the end of his life, filmm...  Dick Johnson Is Dead   
9     A woman adjusting to life after a loss contend...                   NaN   
15    Students of color navigate the daily slights a...                   NaN   
27    Mourning the loss of their beloved junior high...                   NaN   
28    A family’s idyllic suburban life shatters when...                   NaN   
...                                                 ...                   ...   
8791  When a divorced writer gets a letter from an o...                   NaN   
8793  When a father of eight and a mother of 10 prep...                   NaN   
8802  A political cartoonist, a crime reporter and a...                   NaN   
8804  Looking to survive in a world taken over by zo...                   NaN   
8805  Dragged from civilian life, a former superhero...                   NaN   

            Genre         Premiere  Runtime  IMDB Score Language  
0     Documentary  October 2, 2020     90.0         7.5  

In [122]:
#1.4 Muestro todas las películas (type "Movie") y las columnas title y director.
df_merge.loc[df_merge["type"] == 'Movie', ["title", "director"]]

title                       director
0                 Dick Johnson Is Dead                Kirsten Johnson
6     My Little Pony: A New Generation  Robert Cullen, José Luis Ucha
7                              Sankofa                   Haile Gerima
9                         The Starling                 Theodore Melfi
12                        Je Suis Karl            Christian Schwochow
...                                ...                            ...
8801                           Zinzana                Majid Al Ansari
8802                            Zodiac                  David Fincher
8804                        Zombieland                Ruben Fleischer
8805                              Zoom                   Peter Hewitt
8806                            Zubaan                    Mozez Singh

[6131 rows x 2 columns]

In [132]:
#1.5 Muestro los títulos y géneros de las producciones lanzadas en 2018.
df_merge.loc[df_merge["release_year"] == 2018, ["title", "listed_in"]]

title  \
21                               Resurrection: Ertugrul   
37                                          Angry Birds   
73                                         King of Boys   
94                                            Show Dogs   
96    If I Leave Here Tomorrow: A Film About Lynyrd ...   
...                                                 ...   
8732                                      White Chamber   
8743                                           Wildlife   
8774                                          يوم الدين   
8775                                    Yeh Meri Family   
8803                                        Zombie Dumb   

                                              listed_in  
21    International TV Shows, TV Action & Adventure,...  
37                                Kids' TV, TV Comedies  
73                         Dramas, International Movies  
94                   Children & Family Movies, Comedies  
96                      Documentaries, Music & Musicals  
...                                                 ...  
8732                        Sci-Fi & Fantasy, Thrillers  
8743                         Dramas, Independent Movies  
8774   Dramas, Independent Movies, International Movies  
8775                International TV Shows, TV Comedies  
8803             Kids' TV, Korean TV Shows, TV Comedies  

[1147 rows x 2 columns]

In [136]:
#1.6 Muestro el "title" filas donde la columna "director" esté vacía.

pd.DataFrame(df_merge.loc[pd.isnull(df_merge["director"]), "title"])

title
1                           Blood & Water
3                   Jailbirds New Orleans
4                            Kota Factory
10    Vendetta: Truth, Lies and The Mafia
14        Crime Stories: India Detectives
...                                   ...
8795                      Yu-Gi-Oh! Arc-V
8796                           Yunus Emre
8797                            Zak Storm
8800                   Zindagi Gulzar Hai
8803                          Zombie Dumb

[2634 rows x 1 columns]

In [139]:
# Selecciona las primeras 5 filas del conjunto de datos y muestra solo las columnas title, director y country.

df_merge.iloc[0:5,[2,3,5]]

title         director        country
0   Dick Johnson Is Dead  Kirsten Johnson  United States
1          Blood & Water              NaN   South Africa
2              Ganglands  Julien Leclercq            NaN
3  Jailbirds New Orleans              NaN            NaN
4           Kota Factory              NaN          India

In [141]:
for i,j in enumerate(df_merge.columns):
    print(f'{i}:{j}')
    

0:show_id
1:type
2:title
3:director
4:cast
5:country
6:date_added
7:release_year
8:rating
9:duration
10:listed_in
11:description
12:Title
13:Genre
14:Premiere
15:Runtime
16:IMDB Score
17:Language


In [143]:
# - Selecciona las primeras 10 filas del dataset y muestra las columnas desde la segunda hasta la quinta (incluidas).

df_merge.iloc[0:10,2:6]

title                       director  \
0              Dick Johnson Is Dead                Kirsten Johnson   
1                     Blood & Water                            NaN   
2                         Ganglands                Julien Leclercq   
3             Jailbirds New Orleans                            NaN   
4                      Kota Factory                            NaN   
5                     Midnight Mass                  Mike Flanagan   
6  My Little Pony: A New Generation  Robert Cullen, José Luis Ucha   
7                           Sankofa                   Haile Gerima   
8     The Great British Baking Show                Andy Devonshire   
9                      The Starling                 Theodore Melfi   

                                                cast  \
0                                                NaN   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   
2  Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...   
3                                                NaN   
4  Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...   
5  Kate Siegel, Zach Gilford, Hamish Linklater, H...   
6  Vanessa Hudgens, Kimiko Glenn, James Marsden, ...   
7  Kofi Ghanaba, Oyafunmike Ogunlano, Alexandra D...   
8  Mel Giedroyc, Sue Perkins, Mary Berry, Paul Ho...   
9  Melissa McCarthy, Chris O'Dowd, Kevin Kline, T...   

                                             country  
0                                      United States  
1                                       South Africa  
2                                                NaN  
3                                                NaN  
4                                              India  
5                                                NaN  
6                                                NaN  
7  United States, Ghana, Burkina Faso, United Kin...  
8                                     United Kingdom  
9                                      United States

In [144]:
# Muestra las últimas 7 filas del dataset y selecciona las columnas title, director, y country usando índices posicionales.

df_merge.iloc[-7:,[2,3,5]]

title         director                       country
8800  Zindagi Gulzar Hai              NaN                      Pakistan
8801             Zinzana  Majid Al Ansari  United Arab Emirates, Jordan
8802              Zodiac    David Fincher                 United States
8803         Zombie Dumb              NaN                           NaN
8804          Zombieland  Ruben Fleischer                 United States
8805                Zoom     Peter Hewitt                 United States
8806              Zubaan      Mozez Singh                         India

In [145]:
# Selecciona las filas desde la 15 hasta la 25 (incluidas) y muestra las primeras 4 columnas.

df_merge.iloc[15:26,:5]

show_id     type                                              title  \
15     s16  TV Show                                  Dear White People   
16     s17    Movie  Europe's Most Dangerous Man: Otto Skorzeny in ...   
17     s18  TV Show                                    Falsa identidad   
18     s19    Movie                                          Intrusion   
19     s20  TV Show                                             Jaguar   
20     s21  TV Show    Monsters Inside: The 24 Faces of Billy Milligan   
21     s22  TV Show                             Resurrection: Ertugrul   
22     s23    Movie                                    Avvai Shanmughi   
23     s24    Movie       Go! Go! Cory Carson: Chrissy Takes the Wheel   
24     s25    Movie                                              Jeans   
25     s26  TV Show                               Love on the Spectrum   

                                         director  \
15                                            NaN   
16  Pedro de Echave García, Pablo Azorín Williams   
17                                            NaN   
18                                     Adam Salky   
19                                            NaN   
20                                Olivier Megaton   
21                                            NaN   
22                                 K.S. Ravikumar   
23                        Alex Woo, Stanley Moore   
24                                     S. Shankar   
25                                            NaN   

                                                 cast  
15  Logan Browning, Brandon P. Bell, DeRon Horton,...  
16                                                NaN  
17  Luis Ernesto Franco, Camila Sodi, Sergio Goyri...  
18  Freida Pinto, Logan Marshall-Green, Robert Joh...  
19  Blanca Suárez, Iván Marcos, Óscar Casas, Adriá...  
20                                                NaN  
21  Engin Altan Düzyatan, Serdar Gökhan, Hülya Dar...  
22  Kamal Hassan, Meena, Gemini Ganesan, Heera Raj...  
23  Maisie Benson, Paul Killam, Kerry Gudjohnsen, ...  
24  Prashanth, Aishwarya Rai Bachchan, Sri Lakshmi...  
25                                   Brooke Satchwell

In [147]:
# Muestra las filas 20 a 30 y solo las columnas title, release_year, y rating usando índices posicionales.

df_merge.iloc[20:31,[2,7,8]]

title  release_year rating
20  Monsters Inside: The 24 Faces of Billy Milligan          2021  TV-14
21                           Resurrection: Ertugrul          2018  TV-14
22                                  Avvai Shanmughi          1996  TV-PG
23     Go! Go! Cory Carson: Chrissy Takes the Wheel          2021   TV-Y
24                                            Jeans          1998  TV-14
25                             Love on the Spectrum          2021  TV-14
26                                   Minsara Kanavu          1997  TV-PG
27                                        Grown Ups          2010  PG-13
28                                       Dark Skies          2013  PG-13
29                                         Paranoia          2013  PG-13
30                                  Ankahi Kahaniya          2021  TV-14